src/timely-util/src/operator.rs

```diff
@@ -462,13 +470,37 @@ where
         })
     }
 
-    fn expire_stream_at(&self, expiration: G::Timestamp) -> StreamCore<G, C1> {
-        self.inspect_container(move |data_or_frontier| {
-            if let Err(frontier) = data_or_frontier {
-                assert!(
-                    frontier.is_empty() || AntichainRef::new(frontier).less_than(&expiration),
-                    "frontier {frontier:?} has exceeded expiration {expiration:?}!",
-                );
+    fn expire_stream_at(
+        &self,
+        name: &str,
+        expiration: G::Timestamp,
+        token: Weak<()>,
+    ) -> StreamCore<G, C1> {
+        let name = format!("expire_stream_at({name})");
+        self.unary_frontier(Pipeline, &name.clone(), move |cap, _| {
+            let mut cap = Some(cap.delayed(&expiration));
+            let mut buffer = Default::default();
+            let mut warned = false;
+            move |input, output| {
+                if token.upgrade().is_none() {
+                    drop(cap.take());
+                } else {
+                    let frontier = input.frontier().frontier();
+                    if !frontier.less_than(&expiration) && !warned {
+                        tracing::error!(
+                            name = name,
+                            frontier = ?frontier,
+                            expiration = ?expiration,
+                            "frontier not less than expiration"
+                        );
+                        warned = true;
```

Comment:
Does this imply that we'll allow jumps to say the empty frontier (and just log a warning)? Shouldn't we panic instead to handle cases not caught by the existing `determine_dataflow_expiration` logic?